### DISCUSSION TITLE
Make client side sorting optional

### DISCUSSION THREAD

**Discussion by TobseF** [primefaces/community#1764] [PrimeVue]
The sorting indicator is a nice and easy way to give the user the option for sorting. But it can't be used if the data is already sorted by a backend -Which seems to be the case of many projects in the wild. On paged content the sorting of a single page makes no sense. And the client side sorting make the UI slow gives a bad UX because the user can see the data before it gets sorted for a short time.

## Needed Feature
Make the client side sorting optional. Provide an additional `enableClientSorting` property, with can be set to `false` to disable the sorting in the browser.

![2024-04-25_12h50_06](https://github.com/primefaces/community/assets/1190109/6c297a63-0712-4c95-9ada-63c878622e49)

## How to reproduce
Check out the [Data-Table sample for single sort](https://primevue.org/datatable/#single_sort). 
The fields are sortable by the `sortable` property:
````html
<Column field="code" header="Code" sortable></Column>
````
It's not possible to show the sorting indicator in the header **without** the client side sorting.

## Problem
The problem is, that the client sorts the data in other ways than the database query. For example, the database can use the ordinal of an enum for the sort. The client can only use the text value. In case of an error state label, this is not intended.

For paginated data it makes no sense to just sort the content of the page. For example:
Given 8 numbers paged by size `3` (_[page]_):  
```text
[123][456][78]
       ⬆
````
Page 2 is: `[456]` sorted _asc_ (1 based index).  
But toggle the sort to `desc` on client side would result in:  `[654]`. 
But this is wrong because we need to sort the whole dataset to:
```text
[876][543][21]
       ⬆
```
Then, the correct page content is: `[543]`

In addition loading the table data triggers the sorting after the data is loaded and displayed. So the user just sees the data, sorted by the database, for a short time, before it get's sorted by the DataTable and get's reloaded.

**Comment by slavco86** [comment]
I would suggest instead of a flag implement a `sortFn: (data: T[]) => Promise<T[]>` prop to allow for custom sorting logic to be passed through. 
For @TobseF use case the function could simply be used to trigger server side sorting

**Reply by TobseF** [reply]
If this stops the sorting it would be also solve the problem. In addition it's provides the possibility for enhanced sorting mechanism, which are also on the wishlist. This gives us a good extension point.

There is already an `@sort="onColumnSort"` event which we use to reload the data with the new sort. Is there any mechanism to return or omit this `sort` event to say: "Hey, the table is already sorted, stay calm"?
To work together, the `sortFn` needs to trigger the `sort` event afterwards.

**Reply by slavco86** [reply]
I think the event is _probably_ being triggered not by the function but by the clicking on the sort icon, so those are (or should be ) decoupled. Event is just a proxy of the click event and the function is the actual business logic which is being executed after the action took place

**Comment by m-meier** [comment]
Hm.. I might be missing something, but couldn't you just set all columns to sortable = false? https://primevue.org/datatable/#api.column.props.sortable

Judging from the code it might also work to just set sortMode to an invalid value. It checks for "single" or "multiple" and doesn't sort if it is something else.. but that's kinda hacky and might stop working at any point.

**Reply by slavco86** [reply]
This won't work because you are loosing sorting controls with this prop being `false`. The OPs idea is to have the physical controls to trigger server side sorting without the built in sort logic applied to the data set inside the table component

**Reply by m-meier** [reply]
Ah okay, you're right. That would require some kind of support for a serverside sorting mode.

**Reply by TobseF** [reply]
Just updated the description with a _How to reproduce_ section to make it clear. Yes that's the point: Use the sort indicator in the header for the server side sorting and disable the client side sorting.

**Comment by redfox05** [comment]
+1 for this feature, as we currently use the sorting icons in the headers to trigger server side multi column sorting, which works great. However, if we allow the user to save a table layout, including the sort state, then when this layout is loaded, the data is correctly sorted but the headers do not reflect that a specific Sort is active. So being able to pass data that is already sorted, and be able to show in the columns what kind of sorting is applied, would be great. Seems like this should be do-able?

**Comment by vieira-brz** [comment]
The concerns you've raised about client-side sorting, particularly in web applications that handle data sorting on the backend, are indeed valid. In scenarios where data is already sorted server-side or where pagination only displays a subset of the data, client-side sorting can introduce inconsistencies and degrade user experience due to sorting delays or incorrect results.

### Proposed Feature: Optional Client-Side Sorting
Your suggestion to introduce an `enableClientSorting` property is an excellent approach to address these concerns. Here's how this feature could be effectively implemented and communicated within a component library or framework:

#### Implementation

1. **Add `enableClientSorting` Property:**
   - **Default Value:** The property should default to `true`, maintaining the existing behavior for backwards compatibility.
   - **Usage:** When set to `false`, client-side sorting should be disabled. Instead, the sorting indicators in the UI could merely signal to the backend that a sort operation is requested by the user. This would be akin to a "sorting request" rather than performing the sort.

2. **Modify the Column Component:**
   ```html
   <Column field="code" header="Code" sortable enableClientSorting="false"></Column>
   ```

   In this setup, if `enableClientSorting` is `false`, clicking the column's sort header would trigger an event or call a function that could be handled by the backend to sort data accordingly and then fetch the sorted data.

3. **Event Handling:**
   - Define a new event like `onSortRequest` that fires when a column header is clicked and `enableClientSorting` is disabled.
   - The event could pass necessary parameters like the field to sort by and the desired sort direction.

#### Documentation and Developer Guidance
To ensure developers can implement this feature effectively:
- **Document the new property** clearly in the component's API documentation.
- **Provide examples** showing how to handle sort requests on the backend.
- **Explain the benefits** of disabling client-side sorting when data is paginated or sorted server-side, emphasizing performance benefits and consistency.

#### Use Case Examples
- **Paginated Data:** Illustrate how disabling client-side sorting is crucial when dealing with server-side pagination. Explain that sorting should be done globally on the full dataset, not just the currently visible slice.
- **Performance Optimization:** Highlight scenarios where disabling client-side sorting can prevent unnecessary computations and improve the responsiveness of the application.

**Comment by redfox05** [comment]
Is it just me, or does the reply from vieira-brz seem AI-generated? Theres no reference to Vue, instead generic references to how "[the] feature could be effectively implemented and communicated within a **component library or framework**" Seems odd to me...? Also just the whole way the comment is structured/written.
